### PR TITLE
[SPARK-6231][SQL/DF] Automatically resolve join condition ambiguity for self-joins.

### PR DESCRIPTION
**Pull request by rxin**
See the comment in join function for more information.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 1]
```diff
@@ -171,6 +171,11 @@ case class AttributeReference(
     val exprId: ExprId = NamedExpression.newExprId,
     val qualifiers: Seq[String] = Nil) extends Attribute with trees.LeafNode[Expression] {
 
+  /**
+   * Returns true iff the expression id is the same for both attributes.
```
Comment (damiencarol) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: iff => if ???

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: iff == if and only if

Comment (damiencarol) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: "if and only if" should be much more specific in the method comment IMHO.
iff can be confusing.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: `iff` is [a pretty standard shortening](http://en.wikipedia.org/wiki/If_and_only_if) and is used [throughout the spark code base already](https://github.com/apache/spark/search?utf8=%E2%9C%93&q=iff).

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: I disagree. We use it iff a lot in SQL, and if you Google, the first thing that comes up is "if and only if". Plus this is an internal comment.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by marmbrus** [comment]
Thanks!  Merged to master.

**Comment by marmbrus** [comment]
... and branch 1.4

**Comment by gatorsmile** [comment]
@rxin @marmbrus This fix is unable to resolve the condition ambiguity for nested self join. I also found the self joins could generate incorrect results.

**Pull request closed without merging**